### PR TITLE
fix(executor): Add retry on pods watch to handle timeout.

### LINKED ISSUE #3007: Non-configurable time-out for Kubernetes API causes workflow to fail
Checklist:

 * [x] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [x] I've included the logs.
 
**What happened**:
I am experiencing failures in Argo workflows because the Kubernetes API is timing out. And I can't find a way to override or increase the time-outs because they see to be hard-coded: https://github.com/argoproj/argo/blob/14d58036faa444ee49a4905a632db7e0a5ab60ba/util/retry/retry.go#L16

**What you expected to happen**:
Argo should allow controlling the time-out configuration for Kubernetes API.

**How to reproduce it (as minimally and precisely as possible)**:
On a potentially overloaded Kubernetes cluster, Argo workflows sometimes fail because API requests never respond or do not respond in time.

**Anything else we need to know?**:
Nope

**Environment**:
Baremetal Kubernetes deployed using Kubespray.

- Argo version:
```
$ argo version
argo: v2.7.7+54154c6.dirty
  BuildDate: 2020-05-06T22:37:25Z
  GitCommit: 54154c61eb4fe9f052b04328fb00128568dc20d0
  GitTreeState: dirty
  GitTag: v2.7.7
  GoVersion: go1.13.4
  Compiler: gc
  Platform: darwin/amd64
```

- Kubernetes version :
```
$ kubectl version -o yaml
clientVersion:
  buildDate: "2020-04-16T23:35:15Z"
  compiler: gc
  gitCommit: 52c56ce7a8272c798dbc29846288d7cd9fbae032
  gitTreeState: clean
  gitVersion: v1.18.2
  goVersion: go1.14.2
  major: "1"
  minor: "18"
  platform: darwin/amd64
serverVersion:
  buildDate: "2019-03-25T15:19:22Z"
  compiler: gc
  gitCommit: 2166946f41b36dea2c4626f90a77706f426cdea2
  gitTreeState: clean
  gitVersion: v1.13.5
  goVersion: go1.11.5
  major: "1"
  minor: "13"
  platform: linux/amd64
```

**Other debugging information (if applicable)**:

- workflow result:

```
argo --loglevel DEBUG get jed-temp4-b6csb
Name:                jed-temp4-b6csb
Namespace:           default
ServiceAccount:      argo-deployer
Status:              Succeeded
Created:             Tue May 12 17:13:26 +0200 (15 minutes ago)
Started:             Tue May 12 17:13:26 +0200 (15 minutes ago)
Finished:            Tue May 12 17:14:45 +0200 (13 minutes ago)
Duration:            1 minute 19 seconds
Parameters:          
  pvc-size:          1Ti

STEP                                                                  TEMPLATE   PODNAME                     DURATION  MESSAGE
 ✔ jed-temp4-b6csb                                                    jed-temp4                                        
 ├---✔ create-pv                                                      create-pv                                        
 |   ├-✖ create-pv(0)                                                 create-pv  jed-temp4-b6csb-857857010   41s       Unable to connect to the server: dial tcp 10.233.0.1:443: i/o timeout
 |   └-✔ create-pv(1)                                                 create-pv  jed-temp4-b6csb-1327483247  11s       
 └-·-✔ es-dump(0:eshost:ld60,hour:8,index:5c064ea04d.2020-04-30)(0)   es-dump    jed-temp4-b6csb-3726159225  19s       
   ├-✔ es-dump(1:eshost:ld60,hour:10,index:5c064ea04d.2020-04-30)(0)  es-dump    jed-temp4-b6csb-3517569291  12s       
   └-✔ es-dump(2:eshost:ld60,hour:13,index:5c064ea04d.2020-04-30)(0)  es-dump    jed-temp4-b6csb-2750209939  9s
```

- executor logs:

```
$ kubectl logs jed-temp4-b6csb-857857010 -c init
error: container init is not valid for pod jed-temp4-b6csb-857857010
$ kubectl logs jed-temp4-b6csb-857857010 -c wait
error: container wait is not valid for pod jed-temp4-b6csb-857857010
$ kubectl logs jed-temp4-b6csb-857857010
time="2020-05-12T15:13:37Z" level=info msg="Starting Workflow Executor" version=latest+f952df5.dirty
I0512 15:13:37.227567       1 merged_client_builder.go:122] Using in-cluster configuration
I0512 15:13:37.227791       1 merged_client_builder.go:164] Using in-cluster namespace
time="2020-05-12T15:13:37Z" level=info msg="Creating a docker executor"
time="2020-05-12T15:13:37Z" level=info msg="Executor (version: latest+f952df5.dirty, build_date: 2020-05-12T14:42:40Z) initialized (pod: default/jed-temp4-b6csb-857857010) with template:\n{\"name\":\"create-pv\",\"arguments\":{},\"inputs\":{\"parameters\":[{\"name\":\"pvc-size\",\"value\":\"1Ti\"}]},\"outputs\":{\"parameters\":[{\"name\":\"pvc-name\",\"valueFrom\":{\"jsonPath\":\"{.metadata.name}\"}}]},\"metadata\":{},\"resource\":{\"action\":\"create\",\"manifest\":\"apiVersion: v1\\nkind: PersistentVolumeClaim\\nmetadata:\\n  generateName: jed-temp4-\\n  ownerReferences:\\n  - apiVersion: argoproj.io/v1alpha1\\n    blockOwnerDeletion: true\\n    controller: true\\n    kind: Workflow\\n    name: jed-temp4-b6csb\\n    uid: 21169b15-9463-11ea-a7a9-0cc47a8625d2\\nspec:\\n  accessModes:\\n  - ReadWriteOnce\\n  resources:\\n    requests:\\n      storage: 1Ti\\n  storageClassName: archive-backfill\\n\",\"setOwnerReference\":true},\"retryStrategy\":{\"limit\":3}}"
time="2020-05-12T15:13:37Z" level=info msg="Loading manifest to /tmp/manifest.yaml"
time="2020-05-12T15:13:37Z" level=info msg="kubectl create -f /tmp/manifest.yaml -o json"
time="2020-05-12T15:14:07Z" level=error msg="executor error: Unable to connect to the server: dial tcp 10.233.0.1:443: i/o timeout\ngithub.com/argoproj/argo/errors.New\n\t/go/src/github.com/argoproj/argo/errors/errors.go:49\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).ExecResource\n\t/go/src/github.com/argoproj/argo/workflow/executor/resource.go:36\ngithub.com/argoproj/argo/cmd/argoexec/commands.execResource\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/resource.go:45\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewResourceCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/resource.go:22\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.4-0.20181021141114-fe5e611709b0/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.4-0.20181021141114-fe5e611709b0/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.4-0.20181021141114-fe5e611709b0/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:203\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357"
I0512 15:14:07.310192       1 round_trippers.go:438] PATCH https://10.233.0.1:443/api/v1/namespaces/default/pods/jed-temp4-b6csb-857857010 200 OK in 13 milliseconds
time="2020-05-12T15:14:07Z" level=fatal msg="Unable to connect to the server: dial tcp 10.233.0.1:443: i/o timeout\ngithub.com/argoproj/argo/errors.New\n\t/go/src/github.com/argoproj/argo/errors/errors.go:49\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).ExecResource\n\t/go/src/github.com/argoproj/argo/workflow/executor/resource.go:36\ngithub.com/argoproj/argo/cmd/argoexec/commands.execResource\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/resource.go:45\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewResourceCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/resource.go:22\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.4-0.20181021141114-fe5e611709b0/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.4-0.20181021141114-fe5e611709b0/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/pkg/mod/github.com/spf13/cobra@v0.0.4-0.20181021141114-fe5e611709b0/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:203\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357"
```

- workflow-controller logs:

```
$ kubectl logs -n default $(kubectl get pods -l app=workflow-controller -n default -o name)
...
time="2020-05-12T15:13:26Z" level=info msg="Workflow step group node jed-temp4-b6csb-2767845880 not yet completed" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:13:26Z" level=info msg="Workflow update successful" namespace=default phase=Running resourceVersion=638587512 workflow=jed-temp4-b6csb
time="2020-05-12T15:13:27Z" level=info msg="Processing workflow" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:13:27Z" level=info msg="Updating node jed-temp4-b6csb-857857010 message: ContainerCreating"
time="2020-05-12T15:13:27Z" level=info msg="Workflow step group node jed-temp4-b6csb-2767845880 not yet completed" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:13:27Z" level=info msg="Workflow update successful" namespace=default phase=Running resourceVersion=638587549 workflow=jed-temp4-b6csb
time="2020-05-12T15:13:28Z" level=info msg="Processing workflow" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:13:28Z" level=info msg="Workflow step group node jed-temp4-b6csb-2767845880 not yet completed" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:13:38Z" level=info msg="Processing workflow" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:13:38Z" level=info msg="Updating node jed-temp4-b6csb-857857010 status Pending -> Running"
time="2020-05-12T15:13:38Z" level=info msg="Workflow step group node jed-temp4-b6csb-2767845880 not yet completed" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:13:38Z" level=info msg="Workflow update successful" namespace=default phase=Running resourceVersion=638587885 workflow=jed-temp4-b6csb
time="2020-05-12T15:13:39Z" level=info msg="Processing workflow" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:13:39Z" level=info msg="Workflow step group node jed-temp4-b6csb-2767845880 not yet completed" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:14:07Z" level=info msg="Processing workflow" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:14:07Z" level=info msg="Workflow step group node jed-temp4-b6csb-2767845880 not yet completed" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:14:08Z" level=info msg="Processing workflow" namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:14:08Z" level=info msg="Updating node jed-temp4-b6csb-857857010 status Running -> Failed"
time="2020-05-12T15:14:08Z" level=info msg="Updating node jed-temp4-b6csb-857857010 message: Unable to connect to the server: dial tcp 10.233.0.1:443: i/o timeout"
time="2020-05-12T15:14:08Z" level=info msg="1 child nodes of jed-temp4-b6csb[0].create-pv failed. Trying again..." namespace=default workflow=jed-temp4-b6csb
time="2020-05-12T15:14:08Z" level=info msg="Pod node jed-temp4-b6csb-1327483247 initialized Pending" namespace=default workflow=jed-temp4-b6csb
...
```

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

### PR DESCRIPTION
**Pull request by yoms**
Create a retryer for MainContainerStart.
Add exponential retry for watch pod that can cause issue when using k8sapi executor.
With help of @jsenon

Change after review of #3204

Fixes #3007
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I've signed the CLA.
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo/blob/master/USERS.md).

### CODE REVIEW

File: workflow/executor/executor.go [hunk 1]
```diff
@@ -39,6 +39,13 @@ import (
 	os_specific "github.com/argoproj/argo/workflow/executor/os-specific"
 )
 
+var MainContainerStartRetry = wait.Backoff{
+	Steps:    8,
+	Duration: 1 * time.Second,
+	Factor:   1.0,
+	Jitter:   0.1,
+}
```
Comment (simster7) [inline workflow/executor/executor.go:47]: How did we decide that this was the best backoff?

File: workflow/executor/executor.go [hunk 2]
```diff
@@ -938,8 +945,21 @@ func (we *WorkflowExecutor) waitMainContainerStart() (string, error) {
...
+		err = wait.ExponentialBackoff(MainContainerStartRetry, func() (bool, error) {
+			watchIf, err = podsIf.Watch(opts)
+			if err != nil {
+				log.Debugf("Error retry watching: %v", err)
+				return false, nil
+			}
+			return true, nil
+		})
 		if err != nil {
+			log.Warnf("Exponential retry reached: %v", err)
```
Comment (simster7) [inline workflow/executor/executor.go:962]: I don't think this is necessary here

File: workflow/executor/executor.go [hunk 3]
```diff
@@ -938,8 +945,21 @@ func (we *WorkflowExecutor) waitMainContainerStart() (string, error) {
 		opts := metav1.ListOptions{
 			FieldSelector: fieldSelector.String(),
 		}
-		watchIf, err := podsIf.Watch(opts)
+		log.Infof("Try to start watching")
```
Comment (simster7) [inline workflow/executor/executor.go:948]: Could we remove this?

File: workflow/executor/executor.go [hunk 4]
```diff
@@ -938,8 +945,21 @@ func (we *WorkflowExecutor) waitMainContainerStart() (string, error) {
...
+
+		var err error
+		var watchIf watch.Interface
+
+		err = wait.ExponentialBackoff(MainContainerStartRetry, func() (bool, error) {
+			watchIf, err = podsIf.Watch(opts)
+			if err != nil {
+				log.Debugf("Error retry watching: %v", err)
```
Comment (simster7) [inline workflow/executor/executor.go:956]:
```suggestion
				log.Debugf("Failed to establish watch, retrying: %v", err)
```

File: workflow/executor/executor.go (hunk 1 above)
Comment (yoms) [reply workflow/executor/executor.go:47]: We decide these values after tests on our cluster, at first we wanted to use DefaultRetry which was too low and not configuration.

**Comment by sfc-gh-pkrishnamurthy** [comment]
@yoms Thanks for doing this; there are a few other places that could use the same treatment.
https://github.com/yoms/argo/blob/19fc3aa6d2a1f3d46e7320e8d464b5b0c4dbcaa2/workflow/executor/executor.go#L624 - when attempting to `get pods` . I frequently see 
https://github.com/yoms/argo/blob/19fc3aa6d2a1f3d46e7320e8d464b5b0c4dbcaa2/workflow/executor/executor.go#L925 - when waiting for main container to finish
https://github.com/yoms/argo/blob/19fc3aa6d2a1f3d46e7320e8d464b5b0c4dbcaa2/workflow/executor/executor.go#L1152 - when killing sidecars. Killing more than 1 sidecar fails 100% .

Can the new retryConfig be used in these other places?

**Comment by alexec** [comment]
Available in [v2.10.0-rc7](https://github.com/argoproj/argo/releases/tag/v2.10.0-rc7).

**Pull request merged**